torchgeo/datasets/sentinel.py

```diff
@@ -415,3 +424,47 @@ def plot(
             plt.suptitle(suptitle)
 
         return fig
+
+    def _footprint_from_datasource(
+        self, dataset: DatasetReader | WarpedVRT
+    ) -> MultiPolygon | Polygon:
+        """Extract the true geometric footprint from a Sentinel-2 datasource.
+
+        This method reads the original footprint geometry from the associated
+        Sentinel-2 metadata XML file (`MTD_MSIL1C.xml`) and transforms it from
+        geographic coordinates (EPSG:4326) into the CRS of the given dataset.
+
+        Args:
+            dataset: An open raster dataset, either a :class:`rasterio.io.DatasetReader`
+                or a :class:`rasterio.vrt.WarpedVRT`, from which the footprint will
+                be derived.
+
+        Returns:
+            A :class:`shapely.geometry.Polygon` or
+            :class:`shapely.geometry.MultiPolygon` representing the footprint in
+            the dataset's CRS.
+
+        Raises:
+            KeyError: If the existing metadata file does not contain a ``FOOTPRINT`` tag.
+
+        .. versionadded:: 0.8
+        """
+        if hasattr(dataset, 'src_dataset'):
+            # When dataset is a WarpedVRT
+            filepath = dataset.src_dataset.name
+        else:
+            filepath = dataset.name
+        metadata_path = filepath.split('GRANULE')[0] + 'MTD_MSIL1C.xml'
+        if not os.path.exists(metadata_path):
+            warnings.warn(
+                f'Metadata file not found in Sentinel-2 path: {metadata_path}. '
+                'This is required to extract true footprint for this dataset.'
+                'Defaulting to using raster bounds.'
+            )
```

Comment:
I'm worried that this will be printed hundreds of times for most users. Maybe this should not have a warning message?